cli/config/credentials/oauth_store.go

```diff
@@ -0,0 +1,160 @@
+package credentials
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/docker/cli/cli/config/types"
+	"github.com/docker/cli/cli/oauth"
+	"github.com/docker/docker/registry"
+)
+
+// oauthStore wraps an existing store that transparently handles oauth
+// flows, managing authentication/token refresh and piggybacking off an
+// existing store for storage/retrieval.
+type oauthStore struct {
+	backingStore Store
+	manager      oauth.Manager
+}
+
+// NewOAuthStore creates a new oauthStore backed by the provided store.
+func NewOAuthStore(backingStore Store, manager oauth.Manager) Store {
+	return &oauthStore{
+		backingStore: backingStore,
+		manager:      manager,
+	}
+}
+
+const minimumTokenLifetime = 50 * time.Minute
+
+// Get retrieves the credentials from the backing store, refreshing the
+// access token if the stored credentials are valid for less than minimumTokenLifetime.
+// If the credentials being retrieved are not for the official registry, they are
+// returned as is. If the credentials retrieved do not parse as a token, they are
+// also returned as is.
+func (o *oauthStore) Get(serverAddress string) (types.AuthConfig, error) {
+	if serverAddress != registry.IndexServer {
```

Comment:
Yup! That makes sense, I'll move it for now. I had a very vague idea in my head about this (I wondered why we were passing `isDefaultRegistry` in as a bool instead of checking here and checked commit history) and then forgot to ask about it 😅